File: src/util/symbol_extractor.zig
const std = @import("std");

const c = @cImport({
    @cInclude("tree_sitter/api.h");
    @cInclude("tree_sitter/tree-sitter-go.h");
    @cInclude("tree_sitter/tree-sitter-python.h");
    @cInclude("tree_sitter/tree-sitter-typescript.h");
});

pub const FileSymbols = struct {
    filename: []const u8,
    definitions: [][]const u8,
    calls: [][]const u8,
    imports: [][]const u8,
};

pub const Language = enum { typescript, python, go };

// --- Cache: reuse parser and compiled queries across calls ---

var cached_parser: ?*c.TSParser = null;
var cached_def_queries: [3]?*c.TSQuery = .{ null, null, null };
var cached_call_queries: [3]?*c.TSQuery = .{ null, null, null };
var cached_import_queries: [3]?*c.TSQuery = .{ null, null, null };

fn getParser() ?*c.TSParser {
    if (cached_parser) |p| return p;
    const p = c.ts_parser_new() orelse return null;
    cached_parser = p;
    return p;
}

fn getCachedQuery(cache: *[3]?*c.TSQuery, lang: Language, src: []const u8) ?*c.TSQuery {
    const idx = @intFromEnum(lang);
    if (cache[idx]) |q| return q;
    const grammar = selectGrammar(lang) orelse return null;
    var err_offset: u32 = 0;
    var err_type: c.TSQueryError = c.TSQueryErrorNone;
    const q = c.ts_query_new(grammar, src.ptr, @intCast(src.len), &err_offset, &err_type) orelse return null;
    cache[idx] = q;
    return q;
}

/// Extract function definitions, calls, and imports from a file's patch content.
pub fn extractSymbols(alloc: std.mem.Allocator, filename: []const u8, patch: []const u8) !FileSymbols {
    var buf: [65536]u8 = undefined;
    const source = extractSource(patch, &buf);
    if (source.len == 0) return lexicalFallback(alloc, filename, patch);

    if (detectLanguage(filename)) |lang| {
        return extractWithTreeSitter(alloc, filename, source, lang) catch
            lexicalFallback(alloc, filename, patch);
    }
    return lexicalFallback(alloc, filename, patch);
}

fn extractWithTreeSitter(alloc: std.mem.Allocator, filename: []const u8, source: []const u8, lang: Language) !FileSymbols {
    const parser = getParser() orelse return error.ParserInitFailed;
    const grammar = selectGrammar(lang) orelse return error.GrammarNotFound;
    _ = c.ts_parser_set_language(parser, grammar);

    const tree = c.ts_parser_parse_string(parser, null, source.ptr, @intCast(source.len)) orelse return error.ParseFailed;
    defer c.ts_tree_delete(tree);
    const root = c.ts_tree_root_node(tree);

    var defs: std.ArrayList([]const u8) = .empty;
    var calls_list: std.ArrayList([]const u8) = .empty;
    var imports: std.ArrayList([]const u8) = .empty;

    // Extract definitions
    if (getCachedQuery(&cached_def_queries, lang, selectDefQuery(lang))) |query| {
        try runQuery(alloc, query, root, source, &defs);
    }

    // Extract calls
    if (getCachedQuery(&cached_call_queries, lang, selectCallQuery(lang))) |query| {
        try runQuery(alloc, query, root, source, &calls_list);
    }

    // Extract imports
    if (getCachedQuery(&cached_import_queries, lang, selectImportQuery(lang))) |query| {
        try runQuery(alloc, query, root, source, &imports);
    }

    return .{
        .filename = filename,
        .definitions = try defs.toOwnedSlice(alloc),
        .calls = try calls_list.toOwnedSlice(alloc),
        .imports = try imports.toOwnedSlice(alloc),
    };
}

fn runQuery(alloc: std.mem.Allocator, query: *c.TSQuery, root: c.TSNode, source: []const u8, results: *std.ArrayList([]const u8)) !void {
    const cursor = c.ts_query_cursor_new() orelse return;
    defer c.ts_query_cursor_delete(cursor);
    c.ts_query_cursor_exec(cursor, query, root);

    var match: c.TSQueryMatch = undefined;
    while (c.ts_query_cursor_next_match(cursor, &match)) {
        for (0..match.capture_count) |i| {
            const cap = match.captures[i];
            const start = c.ts_node_start_byte(cap.node);
            const end = c.ts_node_end_byte(cap.node);
            if (end > start and end <= source.len) {
                const name = source[start..end];
                // Deduplicate
                var found = false;
                for (results.items) |existing| {
                    if (std.mem.eql(u8, existing, name)) {
                        found = true;
                        break;
                    }
                }
                if (!found) {
                    try results.append(alloc, try alloc.dupe(u8, name));
                }
            }
        }
    }
}

// --- Language detection ---

pub fn detectLanguage(filename: []const u8) ?Language {
    if (std.mem.endsWith(u8, filename, ".ts") or
        std.mem.endsWith(u8, filename, ".tsx") or
        std.mem.endsWith(u8, filename, ".js") or
        std.mem.endsWith(u8, filename, ".jsx")) return .typescript;
    if (std.mem.endsWith(u8, filename, ".py")) return .python;
    if (std.mem.endsWith(u8, filename, ".go")) return .go;
    return null;
}

// --- Grammar selection ---

fn selectGrammar(lang: Language) ?*const c.TSLanguage {
    return switch (lang) {
        .typescript => c.tree_sitter_typescript(),
        .python => c.tree_sitter_python(),
        .go => c.tree_sitter_go(),
    };
}

// --- Tree-sitter queries ---

fn selectDefQuery(lang: Language) []const u8 {
    return switch (lang) {
        .typescript =>
        \\(function_declaration name: (identifier) @name)
        \\(method_definition name: (property_identifier) @name)
        \\(lexical_declaration (variable_declarator name: (identifier) @name value: (arrow_function)))
        ,
        .python =>
        \\(function_definition name: (identifier) @name)
        ,
        .go =>
        \\(function_declaration name: (identifier) @name)
        \\(method_declaration name: (field_identifier) @name)
        ,
    };
}

fn selectCallQuery(lang: Language) []const u8 {
    return switch (lang) {
        .typescript =>
        \\(call_expression function: (identifier) @name)
        \\(call_expression function: (member_expression property: (property_identifier) @name))
        ,
        .python =>
        \\(call function: (identifier) @name)
        \\(call function: (attribute attribute: (identifier) @name))
        ,
        .go =>
        \\(call_expression function: (identifier) @name)
        \\(call_expression function: (selector_expression field: (field_identifier) @name))
        ,
    };
}

fn selectImportQuery(lang: Language) []const u8 {
    return switch (lang) {
        .typescript =>
        \\(import_statement source: (string (string_fragment) @source))
        \\(call_expression function: (identifier) @_req (#eq? @_req "require") arguments: (arguments (string (string_fragment) @source)))
        ,
        .python =>
        \\(import_from_statement module_name: (dotted_name) @source)
        \\(import_statement name: (dotted_name) @source)
        ,
        .go =>
        \\(import_spec path: (interpreted_string_literal) @source)
        ,
    };
}

// --- Extract source from patch ---

pub fn extractSource(patch: []const u8, buf: []u8) []const u8 {
    var out_len: usize = 0;
    var lines = std.mem.splitScalar(u8, patch, '\n');
    while (lines.next()) |line| {
        const content: []const u8 = blk: {
            if (line.len == 0) break :blk "\n";
            if (std.mem.startsWith(u8, line, "+++") or std.mem.startsWith(u8, line, "---") or std.mem.startsWith(u8, line, "@@")) continue;
            if (line[0] == '+') break :blk line[1..];
            if (line[0] == ' ') break :blk line[1..];
            continue;
        };
        if (out_len + content.len + 1 >= buf.len) break;
        @memcpy(buf[out_len .. out_len + content.len], content);
        out_len += content.len;
        buf[out_len] = '\n';
        out_len += 1;
    }
    return buf[0..out_len];
}

// --- Lexical fallback for unsupported languages ---

fn lexicalFallback(alloc: std.mem.Allocator, filename: []const u8, patch: []const u8) !FileSymbols {
    var defs: std.ArrayList([]const u8) = .empty;
    var calls_list: std.ArrayList([]const u8) = .empty;

    var buf: [65536]u8 = undefined;
    const source = extractSource(patch, &buf);

    // Scan for function definitions: function/def/func/fn followed by identifier
    var i: usize = 0;
    while (i < source.len) : (i += 1) {
        if (matchKeyword(source, i, "function") orelse
            matchKeyword(source, i, "def") orelse
            matchKeyword(source, i, "func") orelse
            matchKeyword(source, i, "fn")) |end|
        {
            if (extractIdentifier(source, end)) |ident| {
                const name = try alloc.dupe(u8, ident);
                if (!contains(defs.items, name)) {
                    try defs.append(alloc, name);
                }
                i = end + ident.len;
                continue;
            }
        }

        // Scan for calls: identifier followed by '('
        if (isIdentStart(source[i])) {
            const start = i;
            while (i < source.len and isIdentChar(source[i])) : (i += 1) {}
            if (i < source.len and source[i] == '(') {
                const name = try alloc.dupe(u8, source[start..i]);
                if (!contains(calls_list.items, name)) {
                    try calls_list.append(alloc, name);
                }
            }
        }
    }

    return .{
        .filename = filename,
        .definitions = try defs.toOwnedSlice(alloc),
        .calls = try calls_list.toOwnedSlice(alloc),
        .imports = try alloc.alloc([]const u8, 0),
    };
}

fn matchKeyword(source: []const u8, pos: usize, keyword: []const u8) ?usize {
    if (pos + keyword.len >= source.len) return null;
    if (!std.mem.eql(u8, source[pos .. pos + keyword.len], keyword)) return null;
    // Must be preceded by newline or start of string
    if (pos > 0 and isIdentChar(source[pos - 1])) return null;
    const end = pos + keyword.len;
    // Must be followed by whitespace
    if (end >= source.len or (source[end] != ' ' and source[end] != '\t')) return null;
    return end;
}

fn extractIdentifier(source: []const u8, pos: usize) ?[]const u8 {
    var start = pos;
    // Skip whitespace
    while (start < source.len and (source[start] == ' ' or source[start] == '\t')) : (start += 1) {}
    if (start >= source.len or !isIdentStart(source[start])) return null;
    var end = start;
    while (end < source.len and isIdentChar(source[end])) : (end += 1) {}
    return source[start..end];
}

fn isIdentStart(ch: u8) bool {
    return (ch >= 'a' and ch <= 'z') or (ch >= 'A' and ch <= 'Z') or ch == '_';
}

fn isIdentChar(ch: u8) bool {
    return isIdentStart(ch) or (ch >= '0' and ch <= '9');
}

fn contains(items: [][]const u8, name: []const u8) bool {
    for (items) |item| {
        if (std.mem.eql(u8, item, name)) return true;
    }
    return false;
}

File: crates/cli/src/util/symbol_extractor.rs
use std::collections::HashSet;

use anyhow::Result;
use streaming_iterator::StreamingIterator;
use tree_sitter::{Language, Parser, Query, QueryCursor};

#[allow(dead_code)]
#[derive(Debug, Clone)]
pub struct FileSymbols {
    pub filename: String,
    pub definitions: Vec<String>,
    pub calls: Vec<String>,
    pub imports: Vec<String>,
    pub identifiers: Vec<String>,
}

#[derive(Debug, Clone, Copy)]
pub enum Lang {
    TypeScript,
    Python,
    Go,
}

pub fn detect_language(filename: &str) -> Option<Lang> {
    if filename.ends_with(".ts")
        || filename.ends_with(".tsx")
        || filename.ends_with(".js")
        || filename.ends_with(".jsx")
    {
        Some(Lang::TypeScript)
    } else if filename.ends_with(".py") {
        Some(Lang::Python)
    } else if filename.ends_with(".go") {
        Some(Lang::Go)
    } else {
        None
    }
}

fn get_language(lang: Lang) -> Language {
    match lang {
        Lang::TypeScript => tree_sitter_typescript::LANGUAGE_TYPESCRIPT.into(),
        Lang::Python => tree_sitter_python::LANGUAGE.into(),
        Lang::Go => tree_sitter_go::LANGUAGE.into(),
    }
}

fn def_query(lang: Lang) -> &'static str {
    match lang {
        Lang::TypeScript => {
            r#"(function_declaration name: (identifier) @name)
(method_definition name: (property_identifier) @name)
(lexical_declaration (variable_declarator name: (identifier) @name value: (arrow_function)))"#
        }
        Lang::Python => "(function_definition name: (identifier) @name)",
        Lang::Go => {
            r#"(function_declaration name: (identifier) @name)
(method_declaration name: (field_identifier) @name)"#
        }
    }
}

fn call_query(lang: Lang) -> &'static str {
    match lang {
        Lang::TypeScript => {
            r#"(call_expression function: (identifier) @name)
(call_expression function: (member_expression property: (property_identifier) @name))"#
        }
        Lang::Python => {
            r#"(call function: (identifier) @name)
(call function: (attribute attribute: (identifier) @name))"#
        }
        Lang::Go => {
            r#"(call_expression function: (identifier) @name)
(call_expression function: (selector_expression field: (field_identifier) @name))"#
        }
    }
}

fn import_query(lang: Lang) -> &'static str {
    match lang {
        Lang::TypeScript => {
            r#"(import_statement source: (string (string_fragment) @source))
(call_expression function: (identifier) @_req (#eq? @_req "require") arguments: (arguments (string (string_fragment) @source)))"#
        }
        Lang::Python => {
            r#"(import_from_statement module_name: (dotted_name) @source)
(import_statement name: (dotted_name) @source)"#
        }
        Lang::Go => r#"(import_spec path: (interpreted_string_literal) @source)"#,
    }
}

fn run_query(
    language: &Language,
    query_src: &str,
    source: &[u8],
    tree: &tree_sitter::Tree,
) -> Result<Vec<String>> {
    let query = Query::new(language, query_src)?;
    let mut cursor = QueryCursor::new();
    let mut matches = cursor.matches(&query, tree.root_node(), source);

    let mut results = Vec::new();
    let mut seen = HashSet::new();

    while let Some(m) = matches.next() {
        for cap in m.captures {
            let start = cap.node.start_byte();
            let end = cap.node.end_byte();
            if end > start && end <= source.len() {
                let name = std::str::from_utf8(&source[start..end]).unwrap_or("");
                if !name.is_empty() && seen.insert(name.to_string()) {
                    results.push(name.to_string());
                }
            }
        }
    }

    Ok(results)
}

fn extract_semantic_tokens(source: &str) -> Vec<String> {
    let mut out = Vec::new();
    let mut buf = String::new();
    for ch in source.chars() {
        if ch.is_ascii_alphanumeric() || ch == '_' {
            buf.push(ch.to_ascii_lowercase());
        } else if !buf.is_empty() {
            push_semantic_token(&buf, &mut out);
            buf.clear();
        }
    }
    if !buf.is_empty() {
        push_semantic_token(&buf, &mut out);
    }

    out.sort();
    out.dedup();
    out
}

fn push_semantic_token(token: &str, out: &mut Vec<String>) {
    if token.len() < 6 {
        return;
    }
    if token.bytes().all(|b| b.is_ascii_digit()) {
        return;
    }
    if is_semantic_stopword(token) {
        return;
    }
    out.push(token.to_string());
}

fn is_semantic_stopword(token: &str) -> bool {
    matches!(
        token,
        "import"
            | "export"
            | "default"
            | "const"
            | "let"
            | "function"
            | "return"
            | "class"
            | "interface"
            | "extends"
            | "implements"
            | "public"
            | "private"
            | "protected"
            | "static"
            | "async"
            | "await"
            | "throws"
            | "new"
            | "super"
            | "object"
            | "string"
            | "number"
            | "boolean"
            | "result"
            | "params"
            | "values"
            | "option"
            | "options"
            | "config"
            | "component"
            | "components"
    )
}

/// Extract source code from a unified diff patch.
pub fn extract_source(patch: &str) -> String {
    let mut output = String::new();
    for line in patch.lines() {
        if line.starts_with("+++") || line.starts_with("---") || line.starts_with("@@") {
            continue;
        }
        if line.is_empty() {
            output.push('\n');
            continue;
        }
        match line.as_bytes()[0] {
            b'+' | b' ' => {
                output.push_str(&line[1..]);
                output.push('\n');
            }
            _ => {}
        }
    }
    output
}

/// Extract symbols from a file's patch using tree-sitter or lexical fallback.
pub fn extract_symbols(filename: &str, patch: &str) -> Result<FileSymbols> {
    let source = extract_source(patch);
    if source.is_empty() {
        return Ok(lexical_fallback(filename, patch));
    }

    if let Some(lang) = detect_language(filename) {
        match extract_with_tree_sitter(filename, &source, lang) {
            Ok(symbols) => return Ok(symbols),
            Err(_) => return Ok(lexical_fallback(filename, patch)),
        }
    }
    Ok(lexical_fallback(filename, patch))
}

fn extract_with_tree_sitter(filename: &str, source: &str, lang: Lang) -> Result<FileSymbols> {
    let language = get_language(lang);
    let mut parser = Parser::new();
    parser.set_language(&language)?;

    let tree = parser
        .parse(source.as_bytes(), None)
        .ok_or_else(|| anyhow::anyhow!("parse failed"))?;

    let source_bytes = source.as_bytes();
    let definitions = run_query(&language, def_query(lang), source_bytes, &tree)?;
    let calls = run_query(&language, call_query(lang), source_bytes, &tree)?;
    let imports = run_query(&language, import_query(lang), source_bytes, &tree)?;
    let identifiers = extract_semantic_tokens(source);

    Ok(FileSymbols {
        filename: filename.to_string(),
        definitions,
        calls,
        imports,
        identifiers,
    })
}

fn lexical_fallback(filename: &str, patch: &str) -> FileSymbols {
    let source = extract_source(patch);
    let mut defs = Vec::new();
    let mut calls = Vec::new();
    let mut seen_defs = HashSet::new();
    let mut seen_calls = HashSet::new();
    let identifiers = extract_semantic_tokens(&source);

    let keywords = ["function ", "def ", "func ", "fn "];
    for line in source.lines() {
        let trimmed = line.trim();
        for kw in &keywords {
            if let Some(rest) = trimmed.strip_prefix(kw) {
                if let Some(name) = extract_identifier(rest) {
                    if seen_defs.insert(name.to_string()) {
                        defs.push(name.to_string());
                    }
                }
            }
        }

        // Simple call detection: identifier followed by '('
        let bytes = trimmed.as_bytes();
        let mut i = 0;
        while i < bytes.len() {
            if is_ident_start(bytes[i]) {
                let start = i;
                while i < bytes.len() && is_ident_char(bytes[i]) {
                    i += 1;
                }
                if i < bytes.len() && bytes[i] == b'(' {
                    let name = &trimmed[start..i];
                    if seen_calls.insert(name.to_string()) {
                        calls.push(name.to_string());
                    }
                }
            } else {
                i += 1;
            }
        }
    }

    FileSymbols {
        filename: filename.to_string(),
        definitions: defs,
        calls,
        imports: vec![],
        identifiers,
    }
}

fn extract_identifier(s: &str) -> Option<&str> {
    let s = s.trim_start();
    if s.is_empty() || !is_ident_start(s.as_bytes()[0]) {
        return None;
    }
    let end = s.bytes().position(|b| !is_ident_char(b)).unwrap_or(s.len());
    Some(&s[..end])
}

fn is_ident_start(b: u8) -> bool {
    b.is_ascii_alphabetic() || b == b'_'
}

fn is_ident_char(b: u8) -> bool {
    b.is_ascii_alphanumeric() || b == b'_'
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn extract_source_from_patch() {
        let patch = "@@ -1,3 +1,3 @@\n-old line\n+new line\n context line";
        let source = extract_source(patch);
        assert!(source.contains("new line"));
        assert!(source.contains("context line"));
        assert!(!source.contains("old line"));
    }

    #[test]
    fn detect_typescript() {
        assert!(matches!(detect_language("app.ts"), Some(Lang::TypeScript)));
        assert!(matches!(detect_language("app.tsx"), Some(Lang::TypeScript)));
    }

    #[test]
    fn detect_python() {
        assert!(matches!(detect_language("app.py"), Some(Lang::Python)));
    }

    #[test]
    fn detect_go() {
        assert!(matches!(detect_language("main.go"), Some(Lang::Go)));
    }

    #[test]
    fn detect_unknown() {
        assert!(detect_language("main.rs").is_none());
        assert!(detect_language("Makefile").is_none());
    }

    #[test]
    fn lexical_fallback_finds_definitions() {
        let patch = "@@ -0,0 +1,3 @@\n+function myFunc() {\n+  return 42;\n+}";
        let symbols = lexical_fallback("test.js", patch);
        assert!(symbols.definitions.contains(&"myFunc".to_string()));
    }

    #[test]
    fn lexical_fallback_finds_calls() {
        let patch = "@@ -0,0 +1,1 @@\n+  doSomething(arg);";
        let symbols = lexical_fallback("test.js", patch);
        assert!(symbols.calls.contains(&"doSomething".to_string()));
    }
}
